### PR TITLE
More working commands

### PR DESCRIPTION
**Pull request by AidenMontgomery**
I wanted to get this in front of everyone for feedback on how it is progressing.
I think that there is a lot of room for improvement, but I think that it would be good to get this into the main repo for people to build on.

This is a video of how it is working.
https://cdn.discordapp.com/attachments/903705539857489930/906955032891633735/MoreCommands.mp4

### CODE REVIEW

**Comment by AidenMontgomery** [comment]
Added another commit that performs some checking for existing records with matching details.

**Comment by AidenMontgomery** [comment]
I have rebased after PR #5 was merged.

**Comment by AidenMontgomery** [comment]
Sorry, I keep adding to the PR because I am working on top of it.
I have just pushed a change that will allow the user to specify either a twitter URL or twitter handle and the bot will work out the right format to pass on to AirTable.

**Comment by thenoahhein** [comment]
Ah, glad I read this, I'm looking at an outdated branch then. I'll confirm all the commands work and merge this in

File: commands/addContributor.ts [hunk 1]
```diff
@@ -1,15 +1,44 @@
...
 export async function execute(interaction: CommandInteraction) {
-    if (await isContributor(interaction.user.discriminator)) {
-        interaction.reply('Sorry! You can not add yourself because you are already a contributor!')
-        return
+  if (await isContributor(interaction.user)) {
+    interaction.reply({ content: 'Sorry! You can not add yourself because you are already a contributor!', ephemeral: true });
+    return;
+  }
+
+  const devDAOID = interaction.options.getString('devdao-id');
```
Comment (thenoahhein) [inline commands/addContributor.ts:25]: Would like to do some sanitizing here. In the description we say `eg #2468` but if they put the pound in, it doesn't get submitted to the airtable properly.

File: commands/addResource.ts [hunk 2]
```diff
@@ -1,56 +1,270 @@
...
+    if (contributorResponse) {
+      contributor = contributorResponse.getId();
+    }
+  }
+
+  const url = interaction.options.getString('url');
+  if (!url) {
+    interaction.reply({ content: 'URL required, please try submitting again.', ephemeral: true });
+    return;
+  }
```
Comment (thenoahhein) [inline commands/addResource.ts:107]: Is this just to make typescript happy? With the `isRequired` field being true I would think this check isn't necessary.

File: commands/addResource.ts [hunk 3]
```diff
@@ -1,56 +1,270 @@
...
+
+  const blockchain = await readBlockchain();
+  const bcOptions = blockchain.map(bc => ({ label: bc.name, value: bc.id }));
+  const bcRow = new MessageActionRow().addComponents(
+    new MessageSelectMenu()
+      .setCustomId('blockchain')
+      .setPlaceholder('Select blockchain')
+      .setMaxValues(Math.min(bcOptions.length, 25))
+      .addOptions(bcOptions),
+  );
```
Comment (thenoahhein) [inline commands/addResource.ts]: I'd like to propose renaming it to blockchainOptions to be consistent with the other fields such as `categoryOptions`, although that's certainly non-blocking.

I'd also like to suggest adding an empty field to all of the optional options, as you *have* to answer all of these drop-downs to proceed, but there may be cases that none of the tags are appropriate, and it feels like a bad UX to have to abort all of the progress you made filling out the fields, to have to go back and create a new category/tag/whatever and re-do the process.

This is just my first reaction, happy to discuss further.

File: commands/addResource.ts [hunk 4]
```diff
@@ -1,56 +1,270 @@
...
+      .addOptions(authorOptions),
+  );
+
+  const selectionRows = [authorRow, bcRow, categoryRow, tagsRow];
+
+  await interaction.editReply({
+    embeds: [resourceEmbed],
+    content: 'Tell me about the resource',
+    components: selectionRows,
+  });
```
Comment (thenoahhein) [inline commands/addResource.ts:174]: Could we make this ephemeral too?

File: commands/addContributor.ts [hunk 5]
```diff
@@ -1,15 +1,44 @@
...
+  .setName('add-contributor')
+  .setDescription('add yourself to the contributor list')
+  .addStringOption(
+    option => option.setRequired(false)
+      .setName('devdao-id')
+      .setDescription('Enter your DevDAO ID (e.g. #2468)'))
+  .addStringOption(
+    option => option.setRequired(false)
+      .setName('twitter')
+      .setDescription('Enter your twitter handle'))
```
Comment (thenoahhein) [inline commands/addContributor.ts]: could we provide the format for a happy input like we do in the `devdao-id` section?

File: commands/addContributor.ts [hunk 6]
```diff
@@ -1,15 +1,44 @@
...
+  .addStringOption(
+    option => option.setRequired(false)
+      .setName('eth-wallet-address')
+      .setDescription('Enter your ETH wallet address'));
```
Comment (thenoahhein) [inline commands/addContributor.ts]: I had thought about this when we first made the bot, and I'd like to push back on adding this until we can spend cycles on validating that the person owns the wallet by signing a message. I think eventually we'll be using this table as a source of truth for airdrops, and I would hate for people to have mis-input something here, or not have the right wallet address.

Am I overthinking this? Just want to make sure we get this part right.

File: commands/addResource.ts [hunk 7]
```diff
@@ -1,56 +1,270 @@
...
+          const lookupItem = blockchain.find((value) => value.id === v);
+          return lookupItem ?? { name: 'Unknown', id: v };
+        });
+        break;
+      }
+      case 'author': {
+        resource.author = menuInteraction.values.map(v => {
+          const lookupItem = authors.find((value) => value.id === v);
+          return lookupItem ?? { name: 'Unknown', id: v };
+        })[0];
```
Comment (thenoahhein) [inline commands/addResource.ts]: None of the others have this index at the end. I'm a bit confused what this is doing? My understanding was that author only support one entry.

File: commands/addResourceExtended.ts [hunk 8]
```diff
@@ -1,30 +1,52 @@
 import { SlashCommandBuilder } from '@discordjs/builders';
```
Comment (thenoahhein) [inline commands/addResourceExtended.ts]: FWIW my vote is with `addResource` vs this DM style approach.

File: utils/twitterHandle.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+ * // return { isValid: true, handle: 'aidenmontgomery', URL: 'https://www.twitter.com/aidenmontgomery }
+ * @example
+ * createTwitterHandle('http://www.twitter.com/AidenMontgomery');
+ * // return { isValid: true, handle: 'aidenmontgomery', URL: 'https://www.twitter.com/aidenmontgomery }
+ * @example
+ * createTwitterHandle('this is really long and isnt a valid handle');
+ * // return { isValid: false }
+ * @param   {string}                       input  This should be either a full twitter URL e.g. https://www.twitter.com/AidenMontgomery or a valid twitter handle e.g. AidenMontgomery or @AidenMontgomery
+ * @return  {UTwitterHandleResponse}         When valid inputs are entered returns isValid = true along with the handle and the URL for the twitter account.
+ */
```
Comment (thenoahhein) [inline utils/twitterHandle.ts]: Love the documentation here.

**Review by thenoahhein — CHANGES_REQUESTED** [review]
Left some notes, hopefully others will jump in on the discussion!

File: commands/addResource.ts (hunk 2 above)
Comment (AidenMontgomery) [reply commands/addResource.ts:107]: Yeah, the return type from `options.getString()` is `string | null`

File: commands/addResource.ts (hunk 3 above)
Comment (AidenMontgomery) [reply commands/addResource.ts]: Good point. Let me check that, I did have it initially that you didn't have to answer all of them for the option to finish to appear. Then I added in more select menus.

File: commands/addResource.ts (hunk 4 above)
Comment (AidenMontgomery) [reply commands/addResource.ts:174]: When calling `editReply` `ephemeral` isn't a valid option. I think that I missed it off the new `deferred` response. Will look into it.

File: commands/addContributor.ts (hunk 6 above)
Comment (AidenMontgomery) [reply commands/addContributor.ts]: Maybe I was under thinking it :) I just wanted to include anything that was in the table and the forms. Happy to remove this one for now as I agree with your thinking.

File: commands/addContributor.ts (hunk 5 above)
Comment (AidenMontgomery) [reply commands/addContributor.ts]: Definitely

**Review by thenoahhein — CHANGES_REQUESTED** [review]
With the most recent commits I get the following airtable error when trying to add a resource with a `<SKIPPED>` tag in it.

```
AirtableError {
  error: 'INVALID_RECORD_ID',
  message: 'Value "N/A" is not a valid record ID.',
  statusCode: 422
}
```

File: commands/addResourceExtended.ts (hunk 8 above)
Comment (AidenMontgomery) [reply commands/addResourceExtended.ts]: I will remove the code related to the DM style. I also prefer the interaction based approach more. I didn't know how to accomplish it in that manner until @crondinini show us all how with the /add-author flow.

File: commands/addResource.ts (hunk 7 above)
Comment (AidenMontgomery) [reply commands/addResource.ts]: The return from the menuInteraction is always an array of strings. All of the other select menus are multi select and we want arrays of selected items. In the case of the Author we only allow a single select, so grab the first and _only_ entry in the array.

**Comment by AidenMontgomery** [comment]
I have added in a filter to remove those values before sending it off to the Airtable code. I have run through testing all of the commands again.
I also updated the twitter cleaning code to retain the case of the handle as entered.
With the recent Twitter -> Solana wallet registration, their system appears to be case sensitive!

File: commands/addResourceExtended.ts (hunk 8 above)
Comment (thenoahhein) [reply commands/addResourceExtended.ts]: Shoutout to @crondinini :D

**Comment by thenoahhein** [comment]
Changed one quick thing in regards to the naming table, as there wasn't a space in `discordid` but there is in the table. everything else seems to be running great, and I'm gonna merge it in. Thanks so much, this is really kickass work.

**Pull request merged**